pkg/kubelet/container/helpers.go

```diff
@@ -86,10 +86,17 @@ func ShouldContainerBeRestarted(container *v1.Container, pod *v1.Pod, podStatus
 	if status.State == ContainerStateRunning {
 		return false
 	}
-	// Always restart container in the unknown, or in the created state.
-	if status.State == ContainerStateUnknown || status.State == ContainerStateCreated {
+	// Always restart container in the unknown state
+	if status.State == ContainerStateUnknown {
 		return true
 	}
+	if status.State == ContainerStateCreated {
+		klog.V(4).InfoS("Container state is Created, and it will be decided whether to restart based on the specific situation.", "pod", klog.KObj(pod), "containerName", container.Name, "restartOnCreatedState", restartOnCreatedState)
+		// we don't retart the container directly as the CREATED event will always be triggered
```

Comment:
/cc @smarterclayton

Reply:
/hold

Before I review in detail, I want to make sure I understand the change - from the description I see the assertion:

> ... for handling ContainerCreatedState: with evented pleg, kubelet can catche the middle state after container is created and before container is started, the container is in Created state. For generic PLEG, a Created State always means the container cannot be started, so Generic Container will start a new one when this happens.

So before with generic pleg, why would PLEG sync not report the Created state?  Can you show me what prevents that from being observed?

The only behavior I could see in kuberuntime_container.go to deal with CREATED specifically is:

```
	if status.State != runtimeapi.ContainerState_CONTAINER_CREATED {
		// If container is not in the created state, we have tried and
		// started the container. Set the StartedAt time.
		cStatus.StartedAt = time.Unix(0, status.StartedAt)
	}
```

Reply:
@sairameshv  traced through the code a while ago for this issue, 

https://github.com/kubernetes/kubernetes/issues/121349#issuecomment-1835574484

Reply:
Is

> But in the case of the Evented Pleg, this podWorkerLoop function does not guarantee the same behavior of GetNewerThan mentioned above as the cache gets updated as and when an event is received.

accurate?  This would be a significant bug in evented PLEG if true, so trying to understand whether this issue was resolved?

Reply:
From that comment,


* So, after receiving the CONTAINER_CREATED event, the podWorkerLoop function gets executed completely leaving the pod phase to be still Pending.
* At this point, even before receiving the CONTAINER_STARTED event, if a SyncLoop RECONCILE happens, the [computePodActions](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L824) function results in the ContainersToStart: [0] action that creats a new container which has been present already.


If the container is created and started very quickly, there is a chance the in case of Evented PLEG the kubelet might try to recreate a container that already exists.

Reply:
Can you describe the equivalent behavior of the pod worker in the GenericPLEG case, using the same order?  It would be helpful if you could show what would have happened in generic pleg and what did happen in evented pleg as a side by side so I can be 100% sure of the failure mode.

There are many reasons that SyncPod can be interrupted, causing the next iteration of the cache to have a container created but not yet running from the previous round even under GenericPleg.  However, it's likely to be rare.  If with the generic pleg SyncPod being interrupted causes a created container to be killed and recreated, that's a bug that justifies this change.  

To answer my question from before, it's ok if GenericPLEG is *unlikely* to return a sequence of events that EventedPLEG always returns.  It's *not* ok if EventedPLEG can return a sequence of events that GenericPLEG (by itself, ignoring pod worker) cannot.

Either way, a change of this scope to kubelet flow requires better unit testing and at minimum a specific node e2e test that causes the error with generic pleg.  If that's really really hard to do, we can consider just having the e2e for evented pleg, but I'm pretty sure we can come up with a way to force this rare condition to happen frequently.

Reply:
I think a good way to reproduce this using Generic PLEG would be, if we artificially introduce a delay in CRI runtime (such as cri-o) after the container is created and before it is set in running state. I can give it a shot, let's see if we are able to see it with Generic PLEG as well.

Reply:
Alright, just adding the delay is not going to be enough to reproduce the issue. What we are looking for is, 

1. Make sure the container is created, but not started 
2. Trigger a syncLoop reconciliation in the kubelet, which would catch that container in CREATED state and force a restart. 


Even though kubelet is using Evented PLEG, it also runs Generic PLEG with [increased period](https://github.com/kubernetes/kubernetes/blob/8ba29970e6739bc2e6c6904df8b29d14646be323/pkg/kubelet/pleg/evented.go#L37) .  So  there is a chance that when a container is in created state and Evented PLEG is waiting for the container STARTED event, a periodically scheduled Generic PLEG relisting could trigger a syncLoop reconciliation.

Reply:
It is not Generic PLEG that's causing the syncLoop reconciliation responsible for alpha job failure. I create a PR with Generic PLEG disabled, https://github.com/kubernetes/kubernetes/pull/122793, but the alpha job still failed for it.

Reply:
that PR only increases the timeout from 5 to 50000 🤔 
```diff
- const globalCacheUpdatePeriod = 5 * time.Second
+ const globalCacheUpdatePeriod = 500000 * time.Second
```

Reply:
LOL, sorry, updated a wrong value for testing. Thanks for pointing out. :)

Reply:
this feature touches several codepaths 

```
grep -r utilfeature.DefaultFeatureGate.Enabled\(features.EventedPLEG\) pkg/
pkg/kubelet/kubelet.go: if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/kubelet.go: if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/container/cache.go: if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/kuberuntime/kuberuntime_manager.go:         if idx == 0 && utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/kuberuntime/kuberuntime_manager.go: if !utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/pleg/generic.go:                            if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
pkg/kubelet/pleg/generic.go:    if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) && isEventedPLEGInUse() && status != nil {
```

Reply:
https://github.com/kubernetes/kubernetes/pull/122795

Reply:
Let's try again https://github.com/kubernetes/kubernetes/pull/122794#issuecomment-1892843671

Reply:
https://github.com/kubernetes/kubernetes/pull/122795#issuecomment-1892954136

Reply:
> The thing is all these jobs aren't even actually using the Evented PLEG feature (even though it is enabled). You need to have a CRI runtime that is generating the required CRI events for Evented PLEG to kick in and get in going.

yeah, but we can not have a feature that enabled crashes the cluster because of a missing feature in the runtime, at least not until we make this a hard dependency

Reply:
@aojea @pacoxu I took a closer look the kubelet logs of these jobs and to my surprise, it seems like the runtime they are using, containerd, is sending the CRI events without any special flag. In my experience with cri-o based jobs, you have to explicitly turn on CRI event generation. 

So this would mean that, those jobs are indeed using Evented PLEG. e.g. Take a look the kubelet logs from [this run](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/122794/pull-kubernetes-e2e-kind-alpha-features/1747199866070634496/artifacts/kind-worker2/kubelet.log) (search for `Evented PLEG:`)  and you would see the kubelet is receiving CRI events. 

At this point we can either accept @pacoxu's changes in this [PR](https://github.com/kubernetes/kubernetes/pull/122778/files) which appear to fix the issue or disable it from [starting](https://github.com/kubernetes/kubernetes/blob/c75a146f2346d21f8c8723b79b524e58eec3bea0/pkg/kubelet/kubelet.go#L1648) it altogether. While the proposed fix is really good, I share the concern raised by @smarterclayton here, https://github.com/kubernetes/kubernetes/pull/122763/files#r1452542504.

Reply:
I don't have the knowledge to review this, check with Clayton please

Reply:
@pacoxu I was wondering if this PR can be closed in favor of https://github.com/kubernetes/kubernetes/pull/122778?

Reply:
> @pacoxu I was wondering if this PR can be closed in favor of #122778?

OK. I did test with that PR to make sure if `IsEventedPLEGInUse()` can work well to replace the FG check logics. It works well until now.